flink-libraries/flink-cep/src/main/java/org/apache/flink/cep/nfa/AfterMatchSkipStrategy.java

```diff
@@ -0,0 +1,276 @@
+import org.apache.flink.api.common.typeutils.TypeSerializerConfigSnapshot;
+import org.apache.flink.api.common.typeutils.UnloadableDummyTypeSerializer;
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.core.memory.DataInputViewStreamWrapper;
+import org.apache.flink.core.memory.DataOutputView;
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+import java.util.List;
+
+
+/**
+ * Indicate the skip strategy after a match process.
+ * <p>There're four kinds of strategies:
+ * SKIP_PAST_LAST_EVENT,
+ * SKIP_TO_NEXT_EVENT,
+ * SKIP_TO_FIRST_<code>PATTERN</code> and
+ * SKIP_TO_LAST_<code>PATTERN</code>
+ * </p>
+ */
+public class AfterMatchSkipStrategy implements Serializable {
+
+	// default strategy
+	SkipStrategy strategy = SkipStrategy.SKIP_TO_NEXT_EVENT;
+
+	// fields
+	String patternName = null;
+
+	public AfterMatchSkipStrategy(){
+		this(SkipStrategy.SKIP_TO_NEXT_EVENT, null);
+	}
+
+	public AfterMatchSkipStrategy(SkipStrategy strategy) {
+		this(strategy, null);
+	}
+
+	public AfterMatchSkipStrategy(SkipStrategy strategy, String patternName) {
+		if (patternName == null && (strategy == SkipStrategy.SKIP_TO_FIRST || strategy == SkipStrategy.SKIP_TO_LAST)) {
+			throw new IllegalArgumentException("the patternName field can not be empty when SkipStrategy is " + strategy);
+		}
+		this.strategy = strategy;
+		this.patternName = patternName;
+	}
+
+	public SkipStrategy getStrategy() {
+		return strategy;
+	}
+
+	public String getPatternName() {
+		return patternName;
+	}
+
+	@Override
+	public String toString() {
+		return "AfterMatchStrategy{" +
+			"strategy=" + strategy +
+			", patternName=" + patternName +
+			'}';
+	}
+
+	/**
+	 * Skip Strategy Enum.
+	 */
+	public enum SkipStrategy{
+		SKIP_TO_NEXT_EVENT,
+		SKIP_PAST_LAST_EVENT,
+		SKIP_TO_FIRST,
+		SKIP_TO_LAST
+	}
+
+	/**
+	 * The {@link TypeSerializerConfigSnapshot} serializer configuration to be stored with the managed state.
+	 */
+	public static class AfterMatchSkipStrategyConfigSnapshot extends CompositeTypeSerializerConfigSnapshot {
+
+		private static final int VERSION = 1;
+
+		/**
+		 * This empty constructor is required for deserializing the configuration.
+		 */
+		public AfterMatchSkipStrategyConfigSnapshot() {
+		}
+
+		public AfterMatchSkipStrategyConfigSnapshot(
+			TypeSerializer<SkipStrategy> enumSerializer,
+			TypeSerializer<String> stringSerializer) {
+
+			super(enumSerializer, stringSerializer);
+		}
+
+		@Override
+		public int getVersion() {
+			return VERSION;
+		}
+	}
+
+	/**
+	 *  A {@link TypeSerializer} for the {@link AfterMatchSkipStrategy}.
+	 */
+	public static class AfterMatchSkipStrategySerializer extends TypeSerializer<AfterMatchSkipStrategy> {
```

Comment:
If we remove `AfterMatchSkipStrategy` from `NFA` I think a custom `TypeSerializer` will be unnecessary.